README.md
# Beacon Tool [![CircleCI](https://circleci.com/gh/haensl/beacon-tool.svg?style=svg)](https://circleci.com/gh/haensl/beacon-tool)

Collection of bluetooth beacon signal utility functions.

[![NPM](https://nodei.co/npm/beacon-tool.png?downloads=true)](https://nodei.co/npm/beacon-tool/)

## Features

Beacon Tool allows you to generate, inspect and validate various bluetooth beacon payloads.

## Installation

Via NPM:

`npm i [-g] beacon-tool`

Via Yarn:

`yarn [global] add beacon-tool`

## Usage

### Payload generation

#### `beaconTool.generate(format, [pretty])`

Generates a payload for the given format.

##### `(string, [boolean]) => string`

##### <a name="formats"></a>format `string`

The format to generate.

###### Recognized formats:

* `'ibeacon'`: Generates an ['iBeacon'](https://developer.apple.com/ibeacon/) UUID.
* `'altbeacon'`: Generates an [AltBeacon](https://github.com/AltBeacon/spec) BeaconID.
* `'eddystoneuid'`: Generates an [Eddystone](https://github.com/google/eddystone/tree/master/eddystone-uid) namespace-instance ID pair.

These strings are exposed as constants in [`beaconTool.signals`](#signals).

##### pretty `boolean` *optional*

###### Default: `false`

Whether or not to to add dashes to the returned payload string.

See [`beatify()`](#beatify) for further information.

##### Returns `string`

A payload string for the given result.

#### Example

```javascript
const beaconTool = require('beacon-tool');

const uuid = beaconTool.generate(beaconTool.signals.iBeacon.key);
console.log(uuid);
// de93f975b5d74c9ea8a606fd7cf9d45b

const prettyUUID = beaconTool.generate(beaconTool.signals.iBeacon.key);
console.log(prettyUUID);
// de93f975-b5d7-4c9e-a8a6-06fd7cf9d45b
```

### Payload validation

#### `beaconTool.validate(payload, format)`

Validates a signal payload against a given format.

##### `(string, string) => boolean`

##### payload `string`

The payload to validate.

##### format `string`

The format to generate.

See [Recognized Formats](#formats) for information on format specification.

##### Returns `boolean`

Whether or not the given payload is valid for the given format.

#### Example

```javascript
const beaconTool = require('beacon-tool');

const isValidIBeacon = beaconTool.validate(
  '259771A0-2DFB-4554-B817-2FBFDB5DB1A7',
  beaconTool.signals.iBeacon.key
);

if (isValidIBeacon) {
  console.info('valid');
} else {
  console.info('error');
}
```

### Payload beautifier

#### <a name="#beautify"></a>`beaconTool.beautify(payload, format)`

Beautifies, i.e. adds dashes, to a given payload. `beautify()` returns a new string.

*AltBeacon specification does not specify any representation of it's beacon id with dashes, therefore beautifying an AltBeacon payload will not have any effect.*

##### `(string, string) => string`

##### payload `string`

The payload to beatify.

##### format `string`

The payload format.

See [Recognized Formats](#formats) for information on format specification.

##### Returns `string`

The beatified payload.

#### Example

```javascript
const beaconTool = require('beacon-tool');

const beautifiedPayload = beaconTool.beatify(
  'DE93F975B5D74C9EA8A606FD7CF9D45B',
  beaconTool.signals.iBeacon.key
);

console.log(beatifiedPayload);
// DE93F975-B5D7-4C9E-A8A6-06FD7CF9D45B
```

### Payload identification

#### `beaconTool.identify(payload)`

Tries to identify a given payload.

##### `(string) => array`

#### Example

```javascript
const beaconTool = require('beacon-tool');
const candidates = beaconTool.identify('DE93F975-B5D7-4C9E-A8A6-06FD7CF9D45B');

candidates.forEach((candidate) => {
  console.info(`${beaconTool.get(candidate).displayName}`);
});
// iBeacon
```

##### payload `string`

The payload to identify.

##### Returns `array`

An array of potential candidates. The returned array contains the [`keys`](#signals) of potential candiate formats.

### Payload information

#### `beaconTool.get(format)`

Retrieves information about a given format.

##### `(string) => object`

##### format `string`

The format to retrieve information for.

See [Recognized Formats](#formats) for information on format specification.

##### Returns `object`

A signal information object.

See [signals](#signals) for futher information.

#### Example

```javascript
const beaconTool = require('beacon-tool');
console.log(beaconTool.get('ibeacon'));
/**
{
  dashes: [8, 12, 16, 20],
  displayName: 'iBeacon',
  key: 'ibeacon',
  length: 32
}
*/
```

#### <a name="signals"></a>`beaconTool.signals`

Beacon signal constants.

```javascript
const beaconTool = require('beacon-tool');

console.log(beaconTool.signals);
/**
{
  altBeacon: {
    displayName: 'AltBeacon',
    key: 'altbeacon',
    length: 40
  },
  iBeacon: {
    dashes: [8, 12, 16, 20],
    displayName: 'iBeacon',
    key: 'ibeacon',
    length: 32
  },
  eddystoneUid: {
    dashes: [20],
    displayName: 'Eddystone UID',
    instanceId: {
      length: 12
    },
    key: 'eddystoneuid',
    length: 32,
    namespace: {
      length: 20
    }
  }
}
*/
```

## [Changelog](CHANGELOG.md)

## [License: MIT](LICENCSE)


CHANGELOG.md
## 1.2.7
* [#33: Update dependencies.](https://github.com/haensl/beacon-tool/issues/33)

## 1.2.6
* [#31: Update dependencies.](https://github.com/haensl/beacon-tool/issues/31)

## 1.2.5
* [#29: Add funding information.](https://github.com/haensl/beacon-tool/issues/29)
* Update dependencies.

## 1.2.4
* [#27: Update dependencies.](https://github.com/haensl/beacon-tool/issues/27)

## 1.2.3
* [#25: Update dependencies.](https://github.com/haensl/beacon-tool/issues/25)

## 1.2.2
* [#23: Update dependencies.](https://github.com/haensl/beacon-tool/issues/23)

## 1.2.1
* [#19: Fix build patch.](https://github.com/haensl/beacon-tool/issues/19)

## 1.2.0
* [#17: Migrate to CircleCI.](https://github.com/haensl/beacon-tool/issues/17)

## 1.1.3
* [#15: Update dependencies.](https://github.com/haensl/beacon-tool/issues/15)
* Internalize eslint configuration into `package.json`.

## 1.1.2
* [#11: Update dependencies.](https://github.com/haensl/beacon-tool/issues/11)

## 1.1.1
* [#9: Update dependencies.](https://github.com/haensl/beacon-tool/issues/9)

## 1.1.0
* [#6: Integrate CI deployment.](https://github.com/haensl/beacon-tool/issues/6)

## 1.0.3
* [#5: Update dependencies.](https://github.com/haensl/beacon-tool/issues/5)

## 1.0.2
* [#2: Fix `get()` Eddystone UID](https://github.com/haensl/beacon-tool/issues/2)

## 1.0.1
* Fix get `eddystoneuid`.

## 1.0.0
* Initial project release.


services/signal/signal.test.js
const expect = require('chai').expect;
const signal = require('./');
const rand = require('../../support/random');

const hexRegex = /[a-f0-9]+/;
const alphabet = 'abcdefghijklmnopqrstuvwxyz';
const randString = (len) => {
  let str = '';
  while(str.length < len) {
    str = `${str}${alphabet[rand.randInt(0, alphabet.length - 1)]}`;
  }

  return str;
};

describe('Signal Service', () => {
  describe('generate()', () => {
    describe('iBeacon', () => {
      let uuid;

      beforeEach(() => {
        uuid = signal.generate(signal.signals.iBeacon.key);
      });

      it('generates a hex string', () => {
        expect(hexRegex.test(uuid)).to.be.true;
      });

      it('generates a string of 32 characters', () => {
        expect(uuid.length).to.equal(32);
      });

      describe('pretty', () => {
        beforeEach(() => {
          uuid = signal.generate(signal.signals.iBeacon.key, true);
        });

        it('generates a beautified iBeacon UUID', () => {
          expect(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i.test(uuid)).to.be.true;
        });
      });
    });

    describe('EddystoneUID', () => {
      let uuid;

      beforeEach(() => {
        uuid = signal.generate(signal.signals.eddystoneUid.key);
      });

      it('generates a hex string', () => {
        expect(hexRegex.test(uuid)).to.be.true;
      });

      it('generates a string of 32 characters', () => {
        expect(uuid.length).to.equal(32);
      });

      describe('pretty', () => {
        beforeEach(() => {
          uuid = signal.generate(signal.signals.eddystoneUid.key, true);
        });

        it('generates a beautified Eddystone UUID', () => {
          expect(/^[a-f0-9]{20}-[a-f0-9]{12}$/i.test(uuid)).to.be.true;
        });
      });
    });

    describe('AltBeacon', () => {
      let beaconId;

      beforeEach(() => {
        beaconId = signal.generate(signal.signals.altBeacon.key);
      });

      it('generates a hex string', () => {
        expect(hexRegex.test(beaconId)).to.be.true;
      });

      it('generates a string of 40 characters', () => {
        expect(beaconId.length).to.equal(40);
      });

      describe('pretty', () => {
        beforeEach(() => {
          beaconId = signal.generate(signal.signals.altBeacon.key, true);
        });

        it('generates a beautified AltBeacon BeaconId', () => {
          expect(/^[a-f0-9]{40}$/i.test(beaconId)).to.be.true;
        });
      });
    });

    describe('unknown format', () => {
      it('throws an error', () => {
        expect(() => signal.generate('foo')).to.throw;
      });
    });
  });

  describe('validate', () => {
    describe('iBeacon', () => {
      let uuid;

      describe('valid', () => {
        describe('not pretty', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.iBeacon.key);
          });

          it('returns true', () => {
            expect(signal.validate(uuid, signal.signals.iBeacon.key))
              .to.be.true;
          });
        });

        describe('pretty', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.iBeacon.key, true);
          });

          it('returns true', () => {
            expect(signal.validate(uuid, signal.signals.iBeacon.key))
              .to.be.true;
          });
        });
      });

      describe('invalid', () => {
        describe('length', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.altBeacon.key);
          });

          it('returns false', () => {
            expect(signal.validate(uuid, signal.signals.iBeacon.key))
              .to.be.false;
          });
        });

        describe('charset', () => {
          beforeEach(() => {
            uuid = randString(signal.signals.iBeacon.length);
          });

          it('returns false', () => {
            expect(signal.validate(uuid, signal.signals.iBeacon.key))
              .to.be.false;
          });
        });
      });
    });

    describe('Eddystone-UID', () => {
      let uuid;

      describe('valid', () => {
        describe('not pretty', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.eddystoneUid.key);
          });

          it('returns true', () => {
            expect(signal.validate(uuid, signal.signals.eddystoneUid.key))
              .to.be.true;
          });
        });

        describe('pretty', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.eddystoneUid.key, true);
          });

          it('returns true', () => {
            expect(signal.validate(uuid, signal.signals.eddystoneUid.key))
              .to.be.true;
          });
        });
      });

      describe('invalid', () => {
        describe('length', () => {
          beforeEach(() => {
            uuid = signal.generate(signal.signals.altBeacon.key);
          });

          it('returns false', () => {
            expect(signal.validate(uuid, signal.signals.eddystoneUid.key))
              .to.be.false;
          });
        });

        describe('charset', () => {
          beforeEach(() => {
            uuid = randString(signal.signals.iBeacon.length);
          });

          it('returns false', () => {
            expect(signal.validate(uuid, signal.signals.eddystoneUid.key))
              .to.be.false;
          });
        });
      });
    });

    describe('AltBeacon', () => {
      let beaconId;
      describe('valid', () => {
        beforeEach(() => {
          beaconId = signal.generate(signal.signals.altBeacon.key);
        });

        it('returns true', () => {
          expect(signal.validate(beaconId, signal.signals.altBeacon.key))
            .to.be.true;
        });
      });

      describe('invalid', () => {
        describe('length', () => {
          beforeEach(() => {
            beaconId = signal.generate(signal.signals.iBeacon.key);
          });

          it('returns false', () => {
            expect(signal.validate(beaconId, signal.signals.altBeacon.key))
              .to.be.false;
          });
        });

        describe('charset', () => {
          beforeEach(() => {
            beaconId = randString(signal.signals.iBeacon.length);
          });

          it('returns false', () => {
            expect(signal.validate(beaconId, signal.signals.altBeacon.key))
              .to.be.false;
          });
        });
      });
    });
  });

  describe('beautify()', () => {
    describe('iBeacon', () => {
      let uuid;

      beforeEach(() => {
        uuid = signal.beautify(
          signal.generate(signal.signals.iBeacon.key),
          signal.signals.iBeacon.key
        );
      });

      it('adds dashes at the appropriate places', () => {
        expect(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i.test(uuid)).to.be.true;
      });
    });

    describe('Eddystone UID', () => {
      let uuid;

      beforeEach(() => {
        uuid = signal.beautify(
          signal.generate(signal.signals.eddystoneUid.key),
          signal.signals.eddystoneUid.key
        );
      });

      it('adds dashes at the appropriate places', () => {
        expect(/^[a-f0-9]{20}-[a-f0-9]{12}$/i.test(uuid)).to.be.true;
      });
    });

    describe('AltBeacon', () => {
      let beaconId;

      beforeEach(() => {
        beaconId = signal.beautify(
          signal.generate(signal.signals.altBeacon.key),
          signal.signals.altBeacon.key
        );
      });

      it('does not add any dashes', () => {
        expect(/^[a-f0-9]{40}$/i.test(beaconId)).to.be.true;
      });
    });

    describe('Invalid payload-format combination', () => {
      it('throws', () => {
        expect(() => signal.beautify(
          signal.generate(signal.signals.altBeacon.key),
          signal.signals.iBeacon.key)
        ).to.throw;
      });
    });
  });

  describe('identify', () => {
    let candidates;
    describe('32 chars', () => {
      describe('lowercase, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('de93f975b5d74c9ea8a606fd7cf9d45b');
        });

        it('includes ibeacon', () => {
          expect(candidates).to.include(signal.signals.iBeacon.key);
        });

        it('returns Eddystone UID', () => {
          expect(candidates).to.include(signal.signals.eddystoneUid.key);
        });

        it('returns 2 candidates', () => {
          expect(candidates.length).to.equal(2);
        });
      });

      describe('mixed case, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('dE93F975b5D74c9EA8a606fd7cf9d45b');
        });

        it('includes ibeacon', () => {
          expect(candidates).to.include(signal.signals.iBeacon.key);
        });

        it('returns Eddystone UID', () => {
          expect(candidates).to.include(signal.signals.eddystoneUid.key);
        });

        it('returns 2 candidates', () => {
          expect(candidates.length).to.equal(2);
        });
      });

      describe('uppercase, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('DE93F975B5D74C9EA8A606FD7CF9D45B');
        });

        it('includes ibeacon', () => {
          expect(candidates).to.include(signal.signals.iBeacon.key);
        });

        it('returns Eddystone UID', () => {
          expect(candidates).to.include(signal.signals.eddystoneUid.key);
        });

        it('returns 2 candidates', () => {
          expect(candidates.length).to.equal(2);
        });
      });

      describe('iBeacon dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('DE93F975-B5D7-4C9E-A8A6-06FD7CF9D45B');
        });

        it('includes ibeacon', () => {
          expect(candidates).to.include(signal.signals.iBeacon.key);
        });

        it('returns 1 candidates', () => {
          expect(candidates.length).to.equal(1);
        });
      });

      describe('Eddystone UID dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('DE93F975B5D7-4C9EA8A606FD7CF9D45B');
        });

        it('includes eddystone uid', () => {
          expect(candidates).to.include(signal.signals.eddystoneUid.key);
        });

        it('returns 1 candidates', () => {
          expect(candidates.length).to.equal(1);
        });
      });
    });

    describe('40 characters', () => {
      describe('lowercase, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('de93f975b5d74c9ea8a606fd7cf9d45babcdef01');
        });

        it('includes AltBeacon', () => {
          expect(candidates).to.include(signal.signals.altBeacon.key);
        });

        it('returns 1 candidate', () => {
          expect(candidates.length).to.equal(1);
        });
      });

      describe('mixed case, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('dE93F975b5D74c9EA8a606fd7cf9d45babcdef01');
        });

        it('includes AltBeacon', () => {
          expect(candidates).to.include(signal.signals.altBeacon.key);
        });

        it('returns 1 candidate', () => {
          expect(candidates.length).to.equal(1);
        });
      });

      describe('uppercase, no dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('DE93F975B5D74C9EA8A606FD7CF9D45BABCDEF01');
        });

        it('includes AltBeacon', () => {
          expect(candidates).to.include(signal.signals.altBeacon.key);
        });

        it('returns 1 candidates', () => {
          expect(candidates.length).to.equal(1);
        });
      });

      describe('dashes', () => {
        beforeEach(() => {
          candidates = signal.identify('DE93F975B-5D74C9EA8A606FD7CF-9D45BABCDEF01');
        });

        it('returns no candidates', () => {
          expect(candidates.length).to.equal(0);
        });
      });
    });

    describe('10 characters', () => {
      beforeEach(() => {
        candidates = signal.identify('ADE93F975B');
      });

      it('returns no candidates', () => {
        expect(candidates.length).to.equal(0);
      });
    });

    describe('50 characters', () => {
      beforeEach(() => {
        candidates = signal.identify('ADE93F975BADE93F975BADE93F975BADE93F975BADE93F975B');
      });

      it('returns no candidates', () => {
        expect(candidates.length).to.equal(0);
      });
    });
  });
});


lib/support/random/random.test.js
const expect = require('chai').expect;
const rand = require('./');

describe('random', () => {
  describe('randInt()', () => {
    it('returns a number', () => {
      expect(typeof rand.randInt()).to.equal('number');
    });

    describe('min', () => {
      it('respects the minimum', () => {
        expect(rand.randInt(10)).to.be.least(10);
      });
    });

    describe('max', () => {
      let r;
      beforeEach(() => {
        r = rand.randInt(10, 11);
      });

      it('respects the minimum', () => {
        expect(r).to.be.least(10);
      });

      it('respects the maximum', () => {
        expect(r).to.be.below(11);
      });
    });
  });

  describe('randHex()', () => {
    it('returns a string', () => {
      expect(typeof rand.randHex()).to.equal('string');
    });

    it('returns a hexadecimal character', () => {
      expect(/^[0-f0-9]$/.test(rand.randHex())).to.be.true;
    });
  });
});


signals.js
const hexAlphabet = 'abcdef0123456789'.split('');
const randInt = (min, max) =>
  Math.floor(Math.random() * (Math.floor(max) - Math.floor(min))) + min;
const randHex = () =>
  hexAlphabet[randInt(0, hexAlphabet.length - 1)];

const generateAltBeaconId = () => {
  let beaconId = '';
  let len = 40;
  while (len--) {
    beaconId = `${beaconId}${randHex()}`;
  }

  return beaconId;
};

const generateEddystoneUid = () => {
  let namespace = '';
  let instanceId = '';
  let namespaceLen = 20;
  let instanceIdLen = 12;
  while (namespaceLen--) {
    namespace = `${namespace}${randHex()}`;
  }

  while (instanceIdLen--) {
    instanceId = `${instanceId}${randHex()}`;
  }

  return `${namespace}${instanceId}`;
};

const generateiBeaconUUID = () => {
  let uuid = '';
  let len = 32;
  while (len--) {
    uuid = `${uuid}${randHex()}`;
  }

  return uuid;
};

module.exports = {
  generateiBeaconUUID,
  generateEddystoneUid,
  generateAltBeaconId
};



lib/services/signal/index.js
const rand = require('../../support/random');
const hexRegex = /^[a-f0-9]+$/i;
const hexAndDashesRegex = /^[a-f0-9-]+$/i;

const signals = {
  altBeacon: {
    displayName: 'AltBeacon',
    key: 'altbeacon',
    length: 40,
    regex: /^[a-f0-9]{40}$/i
  },
  iBeacon: {
    dashes: [8, 12, 16, 20],
    displayName: 'iBeacon',
    key: 'ibeacon',
    length: 32,
    regex: /^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i
  },
  eddystoneUid: {
    dashes: [20],
    displayName: 'Eddystone UID',
    instanceId: {
      length: 12
    },
    key: 'eddystoneuid',
    length: 32,
    namespace: {
      length: 20
    },
    regex: /^[a-f0-9]{12}-[a-f0-9]{20}$/i
  }
};

const getLengthForFormat = (format) => {
  let length;
  switch (format.toLowerCase()) {
    case signals.iBeacon.key:
      length = signals.iBeacon.length;
      break;
    case signals.altBeacon.key:
      length = signals.altBeacon.length;
      break;
    case signals.eddystoneUid.key:
      length = signals.eddystoneUid.namespace.length
        + signals.eddystoneUid.instanceId.length;
      break;
    default:
      throw new Error(`Unknown format: ${format}`);
  }

  return length;
};

const beautify = (payload, format) => {
  let beautifiedPayload = `${payload}`;
  let dashes = [];
  switch (format.toLowerCase()) {
    case signals.iBeacon.key:
      dashes = dashes.concat(signals.iBeacon.dashes);
      break;
    case signals.altBeacon.key:
      break;
    case signals.eddystoneUid.key:
      dashes = dashes.concat(signals.eddystoneUid.dashes);
      break;
    default:
      throw new Error(`Unknown format: ${format}`);
  }

  if (dashes.length) {
    beautifiedPayload = dashes.reduce((slices, dashIndex, i, dashes) => {
      slices.push(beautifiedPayload.slice(dashes[i - 1] || 0, dashIndex));
      return slices;
    }, []).concat([ beautifiedPayload.slice(dashes[dashes.length - 1])]).join('-');
  }

  return beautifiedPayload;
};

const generate = (format, pretty = false) => {
  let payload = '';

  while (payload.length < getLengthForFormat(format)) {
    payload = `${payload}${rand.randHex()}`;
  }

  if (pretty) {
    return beautify(payload, format);
  }

  return payload.toUpperCase();
};

const validate = (payload, format) => {
  const normalizedPayload = payload.replace(/-/g, '');
  return hexRegex.test(normalizedPayload)
    && normalizedPayload.length === getLengthForFormat(format);
};

const identify = (payload) => {
  if (!hexAndDashesRegex.test(payload)) {
    return [];
  }

  const knownSignals = [
    signals.iBeacon,
    signals.altBeacon,
    signals.eddystoneUid
  ];
  let candidates;

  if (/-/.test(payload)) {
    candidates = knownSignals.filter(
      (signal) => signal.regex.test(payload)
    );
  } else {
    candidates = knownSignals.filter(
      (signal) => signal.length === payload.length
    );
  }

  return candidates.map((signal) => signal.key);
};

const get = (format) => {
  switch (format) {
    case signals.iBeacon.key:
      return signals.iBeacon;
    case signals.altBeacon.key:
      return signals.altBeacon;
    case signals.eddystoneUid.key:
      return signals.eddystoneUid;
    default:
      throw new Error(`Unknown format: ${format}`);
  }
};

module.exports = {
  beautify,
  generate,
  get,
  identify,
  signals,
  validate
};

